man/unihexgen.1
.TH UNIHEXGEN 1 "2013 Sep 03"
.SH NAME
unihexgen \- Generate Unifont 4\- or 6\-digit hexadecimal glyphs
.SH SYNOPSIS
.br
.B unihexgen
.I startpoint endpoint
.SH DESCRIPTION
.B unihexgen
produces unifont.hex entries in the Unicode code point range
.I startpoint
to
.I endpoint
(specified as the two command\-line arguments).
Output is to STDOUT.  If a codepoint is less than or equal to
"FFFF" (i.e., 0xFFFF), a four\-digit hexadecimal number is encoded
within the corresponding Unifont glyph as two digits on each of
two rows.  Otherwise, a six\-digit hexadecimal number is encoded as
three digits on each of two rows.
.SH OPTIONS
There are no options.
.SH FILES
*.hex as output.
.SH EXAMPLE
To generate the Private Use Area glyphs in the Unicode range
U+E000..U+F8FF, invoke unihexgen with these arguments:
.PP
.RS
.B unihexgen
e000 f8ff
.RE
.SH SEE ALSO
.BR bdfimplode(1),
.BR hex2bdf(1),
.BR hex2sfd(1),
.BR hexbraille(1),
.BR hexdraw(1),
.BR hexkinya(1),
.BR hexmerge(1),
.BR johab2ucs2(1),
.BR unibdf2hex(1),
.BR unibmp2hex(1),
.BR unicoverage(1),
.BR unidup(1),
.BR unifont(5),
.BR unifont\-viewer(1),
.BR unifontchojung(1),
.BR unifontksx(1),
.BR unifontpic(1),
.BR unigencircles(1),
.BR unigenwidth(1),
.BR unihex2bmp(1),
.BR unihex2png(1),
.BR unihexfill(1),
.BR unipagecount(1),
.BR unipng2hex(1)
.SH AUTHOR
.B unihexgen
was written by Paul Hardy.
.SH LICENSE
.B unihexgen
is Copyright \(co 2013 Paul Hardy.
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.SH BUGS
No known real bugs exist, except that this software does not perform
extensive error checking on its command\-line arguments.  If they're not
in the format of the original bitmapped output from
.B unihexgen,
all bets are off.


NEWS
2015-07-02: Release 9.0.01

   Updates to support Unicode 9.0.0.

2015-06-28: Release 8.0.01

  Updates to support Unicode 8.0.0.

2014-10-23: Release 7.0.06

   A few new scripts were added:
      U+010350..U+01037F Old Permic
      U+01F650..U+01F67F Ornamental Dingbats
      U+01F780..U+01F7FF Geometrical Shapes Extended
      U+01F800..U+01F8FF Supplemental Arrows-C

   Various changes were made to Makefiles, Perl scripts, and one bash
   shell script to conform to Debian Policy Manual version 3.9.6.

   Several changes were made to Makefiles:
      * Use "set -e" at the start of multi-command shell invocations,
        to trap errors.
      * Use "&&" rather than ";" in multi-command shell invocations,
        to fail on errors.
      * install now uses "-s" flag for C program binaries.

   Other minor changes were made to scripts; see ChangeLog
   for details.

2014-10-17: Release 7.0.05

   Modified ASCII letter 'l' so it would have the same height as
   the other lower-case letters with ascenders; this is one pixel
   taller than the capital letters.  Lowered the horizontal stroke 
   in 't' by one pixel so it would line up with the horizontal
   stroke in 'f'; this is one pixel lower than the x-height of
   the ASCII letters.  Letters elsewhere in Plane 0 that were formed
   with 'k', 'l', or 't' were checked and modified where appropriate
   to conform with the new ASCII glyphs.

   Completed the Miscellaneous Symbols and Pictographs script in Plane 1,
   and modified Plane 1 Emoticons and Transport and Map Symbols scripts.

   Now create file font/compiled/plane01.html, showing a color-coded
   indication of completion of all of Unicode Plane 1 as an HTML
   table in a self-contained HTML file.  That and the compiled/index.html
   are the two HTML files that building the font creates.  The plane01.html
   file gives a color coding to show completion of each 256-glyph block,
   from 0% (pure red) to 100% (pure green).

2014-10-11: Release 7.0.04

   Added five more Supplemental Multilingual Plane (Plane 1) scripts
   and improved others thanks to several contributors.  There are now
   over 5,000 drawn assigned glyphs for Plane 1!

   Reorganized font/plane01 directory.  Several other changes; see
   ChangeLog file for details.

   unihex2sfd and unipagecount now handle all Unicode planes, not just
   Plane 0.


2014-07-01: Release 7.0.03

   Minor release for Debian.  Added a man page for unihexfill(1);
   Debian requires a man page for each program.

2014-06-22: Release 7.0.02

   Fixes a bug in font/Makefile: there were two "done" statements at the
   end of one "for" loop.  Also adds many new Supplemental Multilingual
   Plane (Plane 1) scripts.  See the ChangeLog file for details.

2014-06-21: Release 7.0.01

   Provides a glyph for every visible code point in the Unicode 7.0 Basic
   Mulitlingual Plane.  Also adds more scripts in the Unicode 7.0
   Supplemental Multilingual Plane.  See the ChangeLog file for full
   details.

   Note that the versioning scheme has changed.  Instead of a date, this
   package will use an incrementing minor version number in the future
   ("01", "02", etc.).

2014-02-14: Release 6.3.20140214

   This version adds the GNU Free Documentation License 1.3 to the "COPYING"
   file.

2014-02-04: Release 6.3.20140204

   This version added Qianqian Fang's name to the COPYRIGHT string now
   passed to Fontforge when the TTF fonts are built, for his contributions
   contained in the file "font/plane00/wqy.hex".

   This version also removes "<img>" links to icons on external websites from
   the "hangul/hangul-generation.html" web page, which were retrieved and
   stored locally when the original web page was saved in the "hangul/"
   directory.

2014-02-02: Release 6.3.20140202

   This version is the first designed to easily build multiple fonts.
   To accomplish this, font/Makefile and font/ttfsrc/Makefile were
   modified to pass FONT and COMBINING parameters, giving the names
   of the font file and its associated combining.txt file.

   This version introduces a subset of Unicode's Supplemental Multilingual
   Plane (SMP, or Plane 1).  The file "font/plane01/plane01.hex" contains
   a glyph for every visible code point in the Unicode 5.0 SMP except
   "Cuneiform" and "Cuneiform Numbers and Punctuation"; those two scripts
   require more resolution than what will fit within a 16 by 16 pixel grid.
   Andrew Miller drew well over 1,000 glyphs for this effort, and added
   glyphs for several SMP scripts introduced after Unicode 5.0.

   This version also introduces a subset of the ConScript Unicode
   Registry (CSUR) Private Use Area glyphs.  Wojciech Stryjewski created
   glyphs for Tengwar and Klingon -- Tolkien fans and Trekkies rejoice!
   Paul Hardy drew Cirth to finish the Tolkien set, and drew other scripts.
   Andrew Miller made very extensive additions for the CSUR; see the
   ChangeLog for details.

   Finally, this version includes a special PSF font for using APL
   in console mode on GNU/Linux.  This PSF font contains 512 glyphs
   as a subset of Unifont; it will not work on systems that only
   support PSF fonts with 256 glyphs.  This PSF font will support
   GNU APL and several other APL implementations.

   The names of the font files are as follows:

      unifont-*               Font with Unifont Plane 0 glyphs

      unifont_sample-*        Font with Unifont glyphs, with combining
                              circles added for illustration

      unifont_csur-*          Font with glyphs from Unifont plus the
                              CSUR PUA glyphs

      unifont_upper-*         Font with glyphs from Planes 1 through 14

      unifont_upper_csur-*    Font with glyphs from Planes 1 through 15

      Unifont-APL8x16.psf.gz  The APL console font

Glyphs for code points U+FFFE and U+FFFF, previously added for illustration
purposes in unifont_sample, were removed because they caused problems on

src/unihexgen.c
/*
   unihexgen.c - generate a series of four- or six-digit hexadecimal
                 numbers within a 16x16 glyph, rendered as white digits
                 on a black background.

                 argv[1] is the starting code point (as a hexadecimal
                 string, with no leading "0x".

                 argv[2] is the ending code point (as a hexadecimal
                 string, with no leading "0x".

                 For example:

                    unihexgen e000 f8ff > pua.hex

                 This generates the Private Use Area glyph file.

   This utility program works in Roman Czyborra's unifont.hex file
   format, the basis of the GNU Unifont package.

   This program is released under the terms of the GNU General Public
   License version 2, or (at your option) a later version.

   Author: Paul Hardy, 2013

   Copyright (C) 2013 Paul Hardy

   LICENSE:

      This program is free software: you can redistribute it and/or modify
      it under the terms of the GNU General Public License as published by
      the Free Software Foundation, either version 2 of the License, or
      (at your option) any later version.
   
      This program is distributed in the hope that it will be useful,
      but WITHOUT ANY WARRANTY; without even the implied warranty of
      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
      GNU General Public License for more details.
   
      You should have received a copy of the GNU General Public License
      along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <stdlib.h>


/*
   hexdigit[][] definition: the bitmap pattern for
   each hexadecimal digit.

   Each digit is drawn as a 4 wide by 5 high bitmap,
   so each digit row is one hexadecimal digit, and
   each entry has 5 rows.

   For example, the entry for digit 1 is:

      {0x2,0x6,0x2,0x2,0x7},

   which corresponds graphically to:

      --#-  ==>  0010  ==>  0x2
      -##-  ==>  0110  ==>  0x6
      --#-  ==>  0010  ==>  0x2
      --#-  ==>  0010  ==>  0x2
      -###  ==>  0111  ==>  0x7

   These row values will then be exclusive-ORed with four one bits
   (binary 1111, or 0xF) to form white digits on a black background.


   The hexdigit[][] array is shared by hexprint4() and hexprint6().
*/
char hexdigit[16][5] = {
   {0x6,0x9,0x9,0x9,0x6},  /* 0x0 */
   {0x2,0x6,0x2,0x2,0x7},  /* 0x1 */
   {0xF,0x1,0xF,0x8,0xF},  /* 0x2 */
   {0xE,0x1,0x7,0x1,0xE},  /* 0x3 */
   {0x9,0x9,0xF,0x1,0x1},  /* 0x4 */
   {0xF,0x8,0xF,0x1,0xF},  /* 0x5 */
   {0x6,0x8,0xE,0x9,0x6},  /* 0x6 */ // {0x8,0x8,0xF,0x9,0xF} [alternate square form of 6]
   {0xF,0x1,0x2,0x4,0x4},  /* 0x7 */
   {0x6,0x9,0x6,0x9,0x6},  /* 0x8 */
   {0x6,0x9,0x7,0x1,0x6},  /* 0x9 */ // {0xF,0x9,0xF,0x1,0x1} [alternate square form of 9]
   {0xF,0x9,0xF,0x9,0x9},  /* 0xA */
   {0xE,0x9,0xE,0x9,0xE},  /* 0xB */
   {0x7,0x8,0x8,0x8,0x7},  /* 0xC */
   {0xE,0x9,0x9,0x9,0xE},  /* 0xD */
   {0xF,0x8,0xE,0x8,0xF},  /* 0xE */
   {0xF,0x8,0xE,0x8,0x8}   /* 0xF */
};


int main(int argc, char *argv[]) {

   int startcp, endcp, thiscp;
   void hexprint4(int); /* function to print one 4-digit unifont.hex code point */
   void hexprint6(int); /* function to print one 6-digit unifont.hex code point */

   if (argc != 3) {
      fprintf(stderr,"\n%s - generate unifont.hex code points as\n", argv[0]);
      fprintf(stderr,"four-digit hexadecimal numbers in a 2 by 2 grid,\n");
      fprintf(stderr,"or six-digit hexadecimal numbers in a 3 by 2 grid.\n");
      fprintf(stderr,"Syntax:\n\n");
      fprintf(stderr,"     %s first_code_point last_code_point > glyphs.hex\n\n", argv[0]);
      fprintf(stderr,"Example (to generate glyphs for the Private Use Area):\n\n");
      fprintf(stderr,"     %s e000 f8ff > pua.hex\n\n", argv[0]);
      exit(EXIT_FAILURE);
   }

   sscanf(argv[1], "%x", &startcp);
   sscanf(argv[2], "%x", &endcp);

   startcp &= 0xFFFFFF; /* limit to 6 hex digits */
   endcp   &= 0xFFFFFF; /* limit to 6 hex digits */

   /*
      For each code point in the desired range, generate a glyph.
   */
   for (thiscp = startcp; thiscp <= endcp; thiscp++) {
      if (thiscp <= 0xFFFF) {
         hexprint4(thiscp); /* print digits 2/line, 2 lines */
      }
      else {
         hexprint6(thiscp); /* print digits 3/line, 2 lines */
      }
   }
   exit(EXIT_SUCCESS);
}


/*
   Takes a 4-digit Unicode code point as an argument
   and prints a unifont.hex string for it to stdout.
*/
void hexprint4(int thiscp) {

   int grid[16]; /* the glyph grid we'll build */

   int row;      /* row number in current glyph */
   int digitrow; /* row number in current hex digit being rendered */
   int rowbits;  /* 1 & 0 bits to draw current glyph row */

   int d1, d2, d3, d4; /* four hexadecimal digits of each code point */

   d1 = (thiscp >> 12) & 0xF;
   d2 = (thiscp >>  8) & 0xF;
   d3 = (thiscp >>  4) & 0xF;
   d4 = (thiscp      ) & 0xF;

   /* top and bottom rows are white */
   grid[0] = grid[15] = 0x0000;

   /* 14 inner rows are 14-pixel wide black lines, centered */
   for (row = 1; row < 15; row++) grid[row] = 0x7FFE;

   printf("%04X:", thiscp);

   /*
      Render the first row of 2 hexadecimal digits
   */
   digitrow = 0; /* start at top of first row of digits to render */
   for (row = 2; row < 7; row++) {
      rowbits = (hexdigit[d1][digitrow] << 9) |
                (hexdigit[d2][digitrow] << 3);
      grid[row] ^= rowbits; /* digits appear as white on black background */
      digitrow++;
   }

   /*
      Render the second row of 2 hexadecimal digits
   */
   digitrow = 0; /* start at top of first row of digits to render */
   for (row = 9; row < 14; row++) {
      rowbits = (hexdigit[d3][digitrow] << 9) |
                (hexdigit[d4][digitrow] << 3);
      grid[row] ^= rowbits; /* digits appear as white on black background */
      digitrow++;
   }

   for (row = 0; row < 16; row++) printf("%04X", grid[row] & 0xFFFF);

   putchar('\n');

   return;
}


/*
   Takes a 6-digit Unicode code point as an argument
   and prints a unifont.hex string for it to stdout.
*/
void hexprint6(int thiscp) {

   int grid[16]; /* the glyph grid we'll build */

   int row;      /* row number in current glyph */
   int digitrow; /* row number in current hex digit being rendered */
   int rowbits;  /* 1 & 0 bits to draw current glyph row */

   int d1, d2, d3, d4, d5, d6; /* six hexadecimal digits of each code point */

   d1 = (thiscp >> 20) & 0xF;
   d2 = (thiscp >> 16) & 0xF;
   d3 = (thiscp >> 12) & 0xF;
   d4 = (thiscp >>  8) & 0xF;
   d5 = (thiscp >>  4) & 0xF;
   d6 = (thiscp      ) & 0xF;

   /* top and bottom rows are white */
   grid[0] = grid[15] = 0x0000;

   /* 14 inner rows are 16-pixel wide black lines, centered */
   for (row = 1; row < 15; row++) grid[row] = 0xFFFF;


   printf("%06X:", thiscp);

   /*
      Render the first row of 3 hexadecimal digits
   */
   digitrow = 0; /* start at top of first row of digits to render */
   for (row = 2; row < 7; row++) {
      rowbits = (hexdigit[d1][digitrow] << 11) |
                (hexdigit[d2][digitrow] <<  6) |
                (hexdigit[d3][digitrow] <<  1);
      grid[row] ^= rowbits; /* digits appear as white on black background */
      digitrow++;
   }

   /*
      Render the second row of 3 hexadecimal digits
   */
   digitrow = 0; /* start at top of first row of digits to render */
   for (row = 9; row < 14; row++) {
      rowbits = (hexdigit[d4][digitrow] << 11) |
                (hexdigit[d5][digitrow] <<  6) |
                (hexdigit[d6][digitrow] <<  1);
      grid[row] ^= rowbits; /* digits appear as white on black background */
      digitrow++;
   }

   for (row = 0; row < 16; row++) printf("%04X", grid[row] & 0xFFFF);

   putchar('\n');

   return;
}

